share/p2p/shrexnd/client.go

```diff
@@ -0,0 +1,143 @@
+package shrexnd
+
+import (
+	"errors"
+	"fmt"
+
+	"time"
+
+	"github.com/ipfs/go-cid"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"golang.org/x/net/context"
+
+	"github.com/celestiaorg/celestia-node/share"
+	"github.com/celestiaorg/celestia-node/share/ipld"
+	pb "github.com/celestiaorg/celestia-node/share/p2p/shrexnd/pb"
+	"github.com/celestiaorg/go-libp2p-messenger/serde"
+	"github.com/celestiaorg/nmt/namespace"
+)
+
+// Client implements client side of shrex/nd protocol to obtain namespaced shares data from remote
+// peers.
+type Client struct {
+	host    host.Host
+	timeout time.Duration
+}
+
+// NewClient creates a new shrEx/nd client
+func NewClient(host host.Host, timeout time.Duration) *Client {
+	return &Client{
+		host:    host,
+		timeout: timeout,
+	}
+}
+
+// GetSharesByNamespace request shares with option to collect proofs from remote peers using shrex
+// protocol
+func (c *Client) GetSharesByNamespace(
+	ctx context.Context,
+	root *share.Root,
+	nID namespace.ID,
+	peerID peer.ID,
+) (share.NamespacedShares, error) {
+	stream, err := c.host.NewStream(ctx, peerID, ndProtocolID)
+	if err != nil {
+		return nil, err
+	}
+	defer stream.Close()
+
+	req := &pb.GetSharesByNamespaceRequest{
+		RootHash:    root.Hash(),
+		NamespaceId: nID,
+	}
+
+	// if context doesn't have deadline use default one
+	deadline, ok := ctx.Deadline()
+	if !ok {
+		deadline = time.Now().Add(c.timeout)
+	}
```

Comment:
We use the pattern where if the client does not pass the timeout through context, it means the client does not want any timeouts. This model makes it obsolete passion the timeout to the client. Opninons?

Reply:
This approach gives:
  - ability for client to set timeout once and and worry about it anymore
  - it is safety feature to prevent unintended unset deadline for client call. There is default value for timeout option, if it is unset. 
  
  I would suggest that If we want to let user to have no timeout, it should be explicit, by setting c.timeout to 0:
  ```
  deadline, ok := ctx.Deadline()
if !ok && c.timeout == 0{
     deadline = time.Now().Add(c.timeout)
}
```